src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,507 @@
+/********************************************************************************
+
+    B-Tree data structure and operations on it.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.util.container.btree.BTree;
+
+import ocean.transition;
+import ocean.core.array.Mutation;
+import ocean.core.array.Search;
+import ocean.core.Enforce;
+import ocean.core.Traits;
+
+import ocean.util.container.mem.MemManager;
+import ocean.util.container.btree.Implementation;
+
+public import ocean.util.container.btree.BTreeRange;
+
+/******************************************************************************
+
+    BTree structure.
+
+
+    BTree is a rooted tree, with the following properties:
+
+    - There's a characteristic of the tree, tree's degree, which indicates
+      how many items are stored in each node. Having a high degree makes the
+      tree shorter, and allows prefetching many elements from the memory in
+      one go.
+
+
+    - The layout of one node of the tree is similar to the following:
+
+      -----------------------------------------------------------------
+      | p1 | element1 | p2 | element2 | ... | pn-1 | element n-1 | pn |
+      -----------------------------------------------------------------
+
+      Every element in the node (for the non-leaf nodes) is surrounded by
+      two pointers. We call these pointers child nodes. Each pointer is a root
+      of a subtree, for which the following stands: all the elements in a tree
+      pointed by p(i-1) (pointer left of the elements) are smaller (in respect
+      to the ordering defined by the element's type) of the element(i), and
+      all the elements in a subtree pointed by p(i) are larger than a element(i).
+
+    - Every node has the following attributes:
+
+        node.number_of_elements = number of elements node contains
```

Comment:
This is actually stored. Remember that the we're allocating nodes as a static arrays. This keeps track of how many elements are _currently_ in the node. (Maximum is always specified as 2*degree-1).